packages/hash/api/src/graph/workspace-types.ts

```diff
@@ -0,0 +1,100 @@
+import {
+  GraphApi,
+  EntityType,
+  PropertyType,
+} from "@hashintel/hash-graph-client";
+import { Logger } from "@hashintel/hash-backend-utils/logger";
+import { AxiosError } from "axios";
+import {
+  DataTypeModel,
+  emailPropertyType,
+  EntityTypeModel,
+  PropertyTypeModel,
+  shortnamePropertyType,
+  userEntityType,
+} from "../model";
+import { logger } from "../logger";
+import {
+  primitiveDataTypeVersionedUris,
+  workspaceAccountId,
+} from "../model/util";
+
+const workspacePropertyTypes: PropertyType[] = [
+  shortnamePropertyType,
+  emailPropertyType,
+];
+
+const workspaceEntityTypes: EntityType[] = [userEntityType];
+
+/**
+ * A script that ensures the required primitive data types and workspace types
+ * have been created in the graph.
+ */
+export const ensureWorkspaceTypesExist = async (params: {
+  graphApi: GraphApi;
+  logger: Logger;
+}) => {
+  const { graphApi } = params;
+
+  // First let's ensure the required data types are in the datastore
+  await Promise.all(
+    Object.entries(primitiveDataTypeVersionedUris).map(
+      async ([title, versionedUri]) => {
+        const dataType = await DataTypeModel.get(graphApi, {
+          versionedUri,
+        }).catch((error: AxiosError) =>
+          error.response?.status === 404 ? null : Promise.reject(error),
+        );
+
+        if (!dataType) {
+          throw new Error(
+            `Primitive data type "${title}" with versioned URI "${versionedUri}" not found`,
+          );
+        }
+      },
+    ),
+  );
+
+  // Next, let's ensure all workspace property types have been created
+  await Promise.all(
+    workspacePropertyTypes.map(async (schema) => {
+      const { $id: versionedUri } = schema;
+      const existingPropertyType = await PropertyTypeModel.get(graphApi, {
```

Comment:
Not something blocking/needed: But since we're either getting or creating the workspace types in these functions, we might be able to cache them to some degree, so that the model classes don't need to fetch every time. It's an optimization though, and not something we need right now

Reply:
> so that the model classes don't need to fetch every time

I'm not sure I'm understanding, where is a type's model class instance fetched multiple times?

Reply:
https://github.com/hashintel/hash/blob/085bc6b4cd42caed3856be654586f77a23036cd1/packages/hash/api/src/model/knowledge/user.model.ts#L98 This model class method is called when the `User` workspace type is needed, but it wouldn't change between calls. But I suppose we want a caching layer anyway, so it might not be big problem

Reply:
I would be inclined to **consciously** try and avoid applying optimisation ideas like this, although we could throw in `@todo`s as long as we're careful to not carelessly implement them later on.

In a similar way to optimisation strategy (not micro-optimising and starting with the big wins), I think we should view caching approaches as the same. This should primarily be solved by making slow requests fast, rather than skipping the requests in the first place. That would give us bigger wins for the app as a whole, and by removing some of those requests prematurely, it might make it difficult for us to actually identify the slowdowns :)

In the long run we could have local caching too of course.

Reply:
Agree with you Ahmad that we need a caching layer to make these calls to the model classes less expensive, but as Alfie says I don't think we should optimise for that now.